### PR TITLE
build: deprecated apollo-federation

### PR DESCRIPTION
**Pull request by aryamohanan**
- [@apollo/federation](https://www.npmjs.com/package/@apollo/federation) is deprecated and  added [@apollo/subgraph](https://www.npmjs.com/package/@apollo/subgraph)
- added support [@apollo/gateway](https://www.npmjs.com/package/@apollo/gateway?activeTab=versions) for 2.x
- Removed graphql v15 from tests and use v16 as default

At present, the default GraphQL version in our repository is v15, which poses challenges during the installation of the @apollo-subgraph and apollo-gateway packages. Both packages mandate GraphQL version 16.5.0 and above.
The decision to maintain GraphQL v15 as the default was driven by the requirement of apollo-express-server v2, which specifically relies on GraphQL v15. To facilitate compatibility with @apollo-subgraph and apollo-gateway, we have opted to eliminate the GraphQL v15 tests from the repository. Upon the availability of v17, we plan to incorporate it into our system.

TODO
- [x] fix failing unit test cases
- [x] fix gql versions

### CODE REVIEW

File: packages/collector/test/tracing/protocols/apollo_subgraph/client.js [hunk 1]
```diff
@@ -10,15 +9,15 @@ require('../../../..')();
 const bodyParser = require('body-parser');
 const express = require('express');
 const morgan = require('morgan');
-const rp = require('request-promise');
+const fetch = require('node-fetch');
```
Comment (aryamohanan) [inline packages/collector/test/tracing/protocols/apollo_subgraph/client.js:12]: request-promise is also a deprecated package, so I changed to node-fetch

File: packages/collector/test/tracing/protocols/apollo_subgraph/gateway.js [hunk 2]
```diff
@@ -31,27 +30,27 @@ if (process.env.WITH_STDOUT) {
 app.use(bodyParser.json());
 
 const gateway = new ApolloGateway({
-  serviceList: [
-    { name: 'accounts', url: `http://localhost:${accountsPort}/graphql` },
-    { name: 'inventory', url: `http://localhost:${inventoryPort}/graphql` },
-    { name: 'products', url: `http://localhost:${productsPort}/graphql` },
-    { name: 'reviews', url: `http://localhost:${reviewsPort}/graphql` }
-  ]
+  supergraphSdl: new IntrospectAndCompose({
```
Comment (aryamohanan) [inline packages/collector/test/tracing/protocols/apollo_subgraph/gateway.js:33]: serviceList also deprecated see, https://www.apollographql.com/docs/federation/v1/gateway/#composing-subgraphs-with-introspectandcompose

File: packages/collector/test/tracing/protocols/apollo_subgraph/client.js (hunk 1 above)
Comment (kirrg001) [reply packages/collector/test/tracing/protocols/apollo_subgraph/client.js:12]: See ticket 150327.

We either do the library replacements in a separate pull request
or you prepare two mergeable commits in this pull request - as you wish :)

**Comment by kirrg001** [comment]
> Removed graphql v15 from tests and use v16 as default

Could you please describe the reasoning? Thanks!

File: packages/collector/test/tracing/protocols/apollo_subgraph/client.js [hunk 3]
```diff
@@ -1,6 +1,5 @@
 /*
- * (c) Copyright IBM Corp. 2021
```
Comment (kirrg001) [inline packages/collector/test/tracing/protocols/apollo_subgraph/client.js:2]: issue: We never change the copyright year in an existing JS file. (even if we rename the file)

File: packages/collector/test/tracing/protocols/graphql/mockVersion.js [hunk 4]
```diff
@@ -1,61 +0,0 @@
-/*
```
Comment (kirrg001) [inline packages/collector/test/tracing/protocols/graphql/mockVersion.js:1]: Q: I am wondering if we want to keep this file. It was really hard to mock a different graphql version.
     As soon as we add graphql v17 (which will come out soon), we need to the file again.

File: packages/collector/test/tracing/protocols/apollo_subgraph/client.js (hunk 3 above)
Comment (aryamohanan) [reply packages/collector/test/tracing/protocols/apollo_subgraph/client.js:2]: Removed.

File: packages/collector/test/tracing/protocols/graphql/test_definition.js [hunk 5]
```diff
@@ -22,18 +22,18 @@ const globalAgent = require('../../../globalAgent');
 
 const agentControls = globalAgent.instance;
 
-function start(graphqlVersion) {
```
Comment (kirrg001) [inline packages/collector/test/tracing/protocols/graphql/test_definition.js:25]: Q: Why did you remove all of the logic? 🤔

**Comment by kirrg001** [comment]
Could you please check the git history which logic we have added for graphql v15 and v16 to make the graphql instrumentation work?

File: package.json [hunk 6]
```diff
@@ -53,8 +53,8 @@
     "depcheck": "lerna exec 'npx depcheck --skip-missing --ignores nan,agent-base,@types/*' --stream"
   },
   "devDependencies": {
-    "@apollo/federation": "^0.38.1",
```
Comment (kirrg001) [inline package.json:56]: Issue: Please add a proper deprecation log to your codebase.

File: package.json [hunk 7]
```diff
@@ -53,8 +53,8 @@
...
-    "@apollo/gateway": "^0.50.2",
```
Comment (kirrg001) [inline package.json:57]: Q: Should `@apollo/gateway` be documented on our public docs? If yes, why? If no, why?

File: packages/collector/test/tracing/protocols/graphql/mockVersion.js (hunk 4 above)
Comment (aryamohanan) [reply packages/collector/test/tracing/protocols/graphql/mockVersion.js]: Restored the file

File: packages/collector/test/tracing/protocols/graphql/test_definition.js (hunk 5 above)
Comment (aryamohanan) [reply packages/collector/test/tracing/protocols/graphql/test_definition.js]: Restored the logic. I missed to consider the future 17 support.

File: package.json (hunk 6 above)
Comment (aryamohanan) [reply package.json:56]: Added

File: packages/collector/test/tracing/protocols/apollo_subgraph/client.js (hunk 1 above)
Comment (aryamohanan) [reply packages/collector/test/tracing/protocols/apollo_subgraph/client.js:12]: I had noticed this on the deprecated list, but during the logic refactoring for Apollo Federation, I opted to remove it from the code. We can handle the deprecation of other usages in a separate pull request.

File: package.json (hunk 7 above)
Comment (aryamohanan) [reply package.json:57]: I am unable to locate any documentation for apollo-gateway in the public documentation. It seems that the majority of the essential tasks related to tracing Apollo Federation gateways (implemented by @apollo/gateway) are handled by our standard GraphQL tracing. The intricate workings of apollo-gateway, which underpins Apollo Federation, might be the reason it is not included in the documentation—this is my speculative assumption.

I propose that we consider adding information about apollo-gateway and subgraph under RPC.

Comment (kirrg001) [reply package.json:57]: > I propose that we consider adding information about apollo-gateway

Yes I think we need to document https://www.npmjs.com/package/@apollo/gateway

File: packages/collector/test/tracing/protocols/apollo_subgraph/test.js [hunk 8]
```diff
@@ -18,16 +17,15 @@ const globalAgent = require('../../../globalAgent');
 const agentControls = globalAgent.instance;
 
 const mochaSuiteFn = supportedVersion(process.versions.node) ? describe : describe.skip;
+// Note: The @apollo/federation package has been deprecated and reached its end-of-life on September 22, 2023.
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/collector/test/tracing/protocols/apollo_subgraph/test.js:20]: updated

**Comment by aryamohanan** [comment]
> Could you please check the git history which logic we have added for graphql v15 and v16 to make the graphql instrumentation work?

Because version graphql v15 is supported by the apollo-express-server v2 and v3 versions, it will function without any problems. Thus, maintaining graphql v15 as the default was not an issue in the past.

File: package.json (hunk 7 above)
Comment (aryamohanan) [reply package.json:57]: I will create a PR for public doc update.

**Comment by kirrg001** [comment]
> Because version graphql v15 is supported by the apollo-express-server v2 and v3 versions, it will function without any problems. Thus, maintaining graphql v15 as the default was not an issue in the past.

Did you go through https://github.com/instana/nodejs/pull/467?

File: package.json [hunk 9]
```diff
@@ -99,8 +99,7 @@
     "admin-plugin-healthcheck": "^1.1.0",
     "amqplib": "^0.10.3",
     "amqplib-v0": "npm:amqplib@^0.8.0",
-    "apollo-server-express": "^2.24.1",
-    "apollo-server-express-v3": "npm:apollo-server-express@^3.6.2",
+    "apollo-server-express": "^3.13.0",
```
Comment (kirrg001) [inline package.json:102]: Q: Why dont we use https://github.com/apollographql/apollo-server/tree/main/packages/server?

File: packages/collector/test/tracing/protocols/apollo_subgraph/gateway.js [hunk 10]
```diff
@@ -1,14 +1,13 @@
...
- * (c) Copyright Instana Inc. and contributors 2020
  */
 
 'use strict';
 
 require('../../../..')();
 
 const { ApolloServer } = require('apollo-server-express');
-const { ApolloGateway } = require('@apollo/gateway');
+const { ApolloGateway, IntrospectAndCompose } = require('@apollo/gateway');
```
Comment (kirrg001) [inline packages/collector/test/tracing/protocols/apollo_subgraph/gateway.js:10]: line 24 needs to be updated I guess

> Apollo Federation Gateway

File: packages/collector/test/tracing/protocols/apollo_subgraph/test.js [hunk 11]
```diff
@@ -18,16 +17,12 @@ const globalAgent = require('../../../globalAgent');
...
+mochaSuiteFn('tracing gateway with apollo-subgraph', function () {
   this.timeout(config.getTestTimeout() * 2);
-
   globalAgent.setUpCleanUpHooks();
-
   const allControls = startAllProcesses();
 
-  [false, true].forEach(withError => registerQuerySuite.bind(this)(allControls, { withError }));
-  // registerQuerySuite.bind(this)({ withError: false });
+  [true, false].forEach(withError => registerQuerySuite.bind(this)(allControls, { withError }));
```
Comment (kirrg001) [inline packages/collector/test/tracing/protocols/apollo_subgraph/test.js:25]: line 90 needs to be updated I guess

>   // Not using ProcessControls.setUpHooks(...) here because it starts all processes simultaneously, but for Apollo
>  // Federation it is necessary to start the processes sequentially and in order.

File: packages/core/src/tracing/instrumentation/protocols/graphql.js [hunk 12]
```diff
@@ -295,6 +295,9 @@ function finishWithException(span, err) {
 }
 
 function instrumentApolloGatewayExecuteQueryPlan(apolloGatewayExecuteQueryPlanModule) {
+  logger.warn(
```
Comment (kirrg001) [inline packages/core/src/tracing/instrumentation/protocols/graphql.js:298]: issue: We only want to log the warning when the customer is using `@apollo/federation`

File: packages/collector/test/tracing/protocols/apollo_subgraph/test.js [hunk 13]
```diff
@@ -18,16 +17,12 @@ const globalAgent = require('../../../globalAgent');
 const agentControls = globalAgent.instance;
 
 const mochaSuiteFn = supportedVersion(process.versions.node) ? describe : describe.skip;
-
-mochaSuiteFn('tracing/apollo-federation', function () {
+mochaSuiteFn('tracing gateway with apollo-subgraph', function () {
   this.timeout(config.getTestTimeout() * 2);
```
Comment (kirrg001) [inline packages/collector/test/tracing/protocols/apollo_subgraph/test.js:21]: Could you pls increase the timeout? The test is failing on my machine. I had to increase it to

> this.timeout(config.getTestTimeout() * 5);

File: packages/core/src/tracing/instrumentation/protocols/graphql.js (hunk 12 above)
Comment (aryamohanan) [reply packages/core/src/tracing/instrumentation/protocols/graphql.js]: done

**Comment by kirrg001** [comment]
Suggestion: rebase all the commits into two commits. One for feature subgraph and one for the deprecation of federation. Then we can merge both commits and they are both visible in the release notes 👍

**Review by kirrg001 — APPROVED** [review]
LGTM 🚀
